tests/codegen/regparm-inreg.rs

```diff
@@ -0,0 +1,44 @@
+// Checks how `regparm` flag works with different calling conventions:
+// marks function arguments as "inreg" like the C/C++ compilers for the platforms.
+// x86 only.
+
+//@ compile-flags: --target i686-unknown-linux-gnu -Zregparm=3 -O -C no-prepopulate-passes
+//@ needs-llvm-components: x86
+
+#![crate_type = "lib"]
+#![no_core]
+#![feature(no_core, lang_items)]
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "copy"]
+trait Copy {}
+
+pub mod tests {
+    // regparm doesn't work for "fastcall" calling conv (only 2 inregs)
+    // CHECK: @f1(i32 inreg noundef %_1, i32 inreg noundef %_2, i32 noundef %_3)
+    #[no_mangle]
+    pub extern "fastcall" fn f1(_: i32, _: i32, _: i32) {}
+
+    // regparm doesn't work for "Rust" calling conv
+    // CHECK: @f2(i32 noundef %_1, i32 noundef %_2, i32 noundef %_3)
+    #[no_mangle]
+    pub extern "Rust" fn f2(_: i32, _: i32, _: i32) {}
```

Comment:
I would have expected that it works for Rust ones too (and also from reading the description of the flag). Or am I misunderstanding this?

Reply:
"Rust" functions don't have any defined ABI, and I don't think we want to have a flag that guarantees things about them.

Reply:
It is not about guaranteeing a particular ABI, but about doing what the flag claims to do.

Reply:
We can't document the flag to do anything for "Rust" functions as that would be stating a guarantee about "Rust" functions.

Reply:
We can definitely document that a flag does (or attempts to do) something without documenting the details of how it accomplishes it, or without guaranteeing stability of those details, or without guaranteeing it will always happen, etc., just like it is done for other things. I don't understand why you claim such a flag can only ever apply to stable ABIs.

Reply:
This would be the first time do any such thing for the Rust ABI (AFAIK). It is risky because you can add as many caveats as you want, people *will* rely on this.

In the end, it's not my call though. I think the reasoning I gave above is why currently the flag does nothing.

Reply:
To be clear, the documentation thing is secondary -- my original point is that there is nothing being done for "Rust" functions here, if I understand correctly the comment there; and what I am saying is that the compiler could at least attempt to do so something, i.e. pass some things in registers. If that is too complicated (implementation-wise), that is fine, but I wanted to understand the reasoning behind not doing anything.

Reply:
> We'd at least need a good self-contained motivation for why this is a good thing to do. If adding more `inreg` makes things faster, shouldn't we always do that?

Yeah, that is a good point, and perhaps `rustc` already marks things as such where it makes sense and, if not, then maybe it should, as you say. I am not sure what is the best trade-off in terms of number of registers, though -- the optimal way of marking the parameters may depend on the function itself, and "Rust" functions could take advantage of that flexibility, i.e. per-function.

I guess it is also conceivable that some projects could want to tweak the number, even if it is just a hint for "Rust" ones. I don't know how popular `-mregparm` and the function attribute are, though, i.e. perhaps nobody actually will ever care for Rust.

Reply:
If we want to give programmers that ability to give hints for the "Rust" ABI, we'd almost certainly want that to be independent from flags that change stable ABIs like "C".

Reply:
> per-function.

That doesn't work, as given a function pointer (i.e., knowing just the signature) we must know how to pass the arguments.

Reply:
> If we want to give programmers that ability to give hints for the "Rust" ABI, we'd almost certainly want that to be independent from flags that change stable ABIs like "C".

Maybe -- I am just talking from the perspective of someone enabling `-mregparm` and what they would expect, i.e. I would expect it affects all codegen (or at least tries to), unless otherwise stated, even if it is not a guarantee for "Rust" functions.

> That doesn't work, as given a function pointer (i.e., knowing just the signature) we must know how to pass the arguments.

That is fine, you can still have flexibility per-signature. I imagine it could also be done per-function for CU-private functions, i.e. anywhere one may be able to guarantee there are no function pointers involved.

Reply:
I think there're two orthogonal things here:
1. Effect on ABI
2. Effect on performance

For (1), it places constraints on "C" ABI only, and should not effect Rust ABI. For (2), Rust compiler, with its freedom on `extern "Rust"`, can freely apply inreg to functions. These are completely orthogonal, it's perfectly fine to enable (2) without (1).

For this reason, I think this PR should focus on (1). The option name probably should also be renamed to reflect that. Maybe `-Zc-abi-reg-param`? (2) is an performance improvement and can be made at any time -- arguably we can just make it a default for all `extern "Rust"` functions on x86 if perf run is positive.

Reply:
-Zregparm was implemented for external calling conv (not for "Rust") because I was not sure if it is need to work for "Rust" too. @ojeda if this flag is good to work for "Rust" calling conv too, it is not big technical problem to support (as far as I see). Should it be done?
@nbdd0121 are you sure that renaming of the flag is a good idea? `regparm` is used in clang/gcc, so it is something like "well-known name".

Reply:
Which ABIs exactly are affected? All non-"Rust" ABIs? Or just "C"?

It would seem strange to have a flag affect *all* ABIs at once when this is really a per-ABI decision.

Reply:
> @nbdd0121 are you sure that renaming of the flag is a good idea? regparm is used in clang/gcc, so it is something like "well-known name".

Well, Rust doesn't blindly follow name chosen by GCC (and clang tries to be CLI compatible so use the same). For example, `-mcmodel` is `-C code-model`. I would suggest choose a name that describes the option best.

> Which ABIs exactly are affected? All non-"Rust" ABIs? Or just "C"?

The code as currently implemented affect `"C"`, `"cdecl"` & `"stdcall"`, i.e. C-ABI and their mappings on x86 Linux/Windows.

Reply:
Added -Zregparm support for native Rust calling conv too. Test improved.